docs/template-resolution-for-dotnet-cli.md

```diff
@@ -0,0 +1,479 @@
+# Template resolution for dotnet CLI
+
+## Template resolution details
+
+In dotnet new CLI, we use template groups in listings instead of individual
+templates.
+
+Template group is set of templates with same `groupIdentity` defined in
+template.json. In case group identity is not set, the template is treated as
+individual template groups.
+
+`dotnet new --list` and `dotnet new --search` show the list of template groups, not
+all the templates.
+
+Templates in template group:
+-   can have different languages (all the languages are shown in ‘Language’
+    column in table listing)
+-   can have different short name: in this case short names can be equally used
+    as synonyms in command input and will be shown in comma separated list in
+    table listing.
+-   assume to have same type, name, author, and tags (classification). When
+    showing the group, the information from the template with highest
+    ‘precedence’ value will be shown; however, the template group will be
+    considered as match if at least one template from the group is a match. For
+    help and instantiating only the templates which match the given type and
+    baseline will be considered.
+
+When resolving the template to be shown/used:
+-   first the template group is evaluated based on short name or name match.
+    Name match/partial matches is only applicable for `--list` and `--search`.
+-   when evaluating the template group, the supported languages are considered:
+    if the template group has multiple languages defined and the user didn’t
+    specify language to use the default language (C\#) will be used, otherwise
+    the input results in error.
+-   then the other matches of the group are evaluated based on the other
+    template info filters (type, author, tag, baseline) and template specific
+    parameters.
+
+In order to instantiate (create) the template, the command input should resolve
+single template from single template group:
+-   in case multiple template groups are resolved, it results in error.
+-   in case of multiple templates within single template group are resolved, the
+    precedence value from template.json will be used to determine the template
+    to run. If the precedence value is not given or is the same for multiple
+    templates, it results in error.
+
+In order to show help for the template, the command input should resolve the
+single template group (based on short name), however, may resolve multiple
+templates, but they should have same language. In case of multiple templates,
```

Comment:
```suggestion
In order to show help for a template, the command input should resolve to a
single template group (based on short name). If the resolved group
contains multiple matching templates, they should have same language. In case of multiple templates,
```